### ISSUE TITLE
Nothing Detected after Training YOLO on my dataset

### ISSUE THREAD

**Issue by ghadahamed** [AlexeyAB/darknet#3358]
@AlexeyAB  first of all sorry for disturbing you but I need your help as I did multiple trials and nothing unfortunately succeed.
I trained on breast images to detect the malignant or benign masses if exist.'
The training set = 86 mammograms
The test set = 21 mammograms
Batch = 1 for training
Batch = 64 & subdivisions = 64 for testing
Width & height = 416
Learning rate = 0.001
The loss still decremented till the iteration 800 and till 1000 decremented with a very small factor till approximately 0.6xxx.
The latest trial, I stopped the training after 1000 iterations and use the generated weights file to used for detection but nothing detected. What is wrong I did or what is missed shall be done to detect any mammogram?

Thanks in advance

**Comment by AlexeyAB** [comment]
@ghadahamed Hi,

> **Batch = 1 for training**
> Batch = 64 & subdivisions = 64 for testing
> Width & height = 416
> Learning rate = 0.001
> The loss still decremented till the iteration 800 and till 1000 decremented with a very small factor till approximately 0.6xxx.

Read this manual, you must train with batch=64: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

Also you should train for at least 2000-4000 iterations and I would recommend you to train with flag `-map` and set `valid=valid.txt` parameter in `obj.data` file to see the mAP chart: https://github.com/AlexeyAB/darknet#when-should-i-stop-training 

What GPU do you use for training?

**Comment by ghadahamed** [comment]
@AlexeyAB Thanks a lot for your help, I started in doing this: change the batch to be 64, valid parameter and add mAP flag.

My GPU is NVIDIA GeForce GTX 960M

I started in training from nearly 3 hours and this is the output till now
![image](https://user-images.githubusercontent.com/47919136/59150654-e19c0680-8a26-11e9-85fe-5068034d127e.png)

**Comment by AlexeyAB** [comment]
Try to use the latest version of Darknet.

**Comment by ghadahamed** [comment]
@AlexeyAB I did this and start training with the following parameters:
# Testing
batch=64
#subdivisions=64
# Training
batch=64
subdivisions=64
learning_rate=0.001
burn_in=1000
max_batches = 4000
policy=steps
steps=3200,3600
scales=.1,.1

This is the output till now
![image](https://user-images.githubusercontent.com/47919136/59167001-40e53e00-8b2f-11e9-9355-aeee1906455b.png)
![image](https://user-images.githubusercontent.com/47919136/59192999-7d8f5480-8b84-11e9-9dc5-34ea685fcf79.png)

Is  this okay? Shall I continue training or there is something wrong?
I am so sorry for disturbing you and thanks a lot

**Comment by AlexeyAB** [comment]
All ok

**Comment by ghadahamed** [comment]
@AlexeyAB 
I am really for disturbing you but unfortunately nothing detected as well neither by using the best weights file or other as at 3000. Even when I test with the same trained image but rotated by different angle, it is not detected as well. While on the other side when test with any trained image, the mass is detected accurately (this is occurs using the weights of different iterations as 1000, 2000, 3000 or best weights) as follows:
TEST USING TRAINED image:
![image](https://user-images.githubusercontent.com/47919136/59386872-0e725580-8d68-11e9-8f96-a840a21fe3d1.png)

TEST USING the same image but flipped (not used in training):
![image](https://user-images.githubusercontent.com/47919136/59386942-3366c880-8d68-11e9-84dc-b4925e24282e.png)

Is there anything wrong I did or something missed before test. 

I put the test data with the trained data in the data/obj folder as well and write its path in the test.txt file in the data folder.

Really sorry for disturbance @AlexeyAB but I tried a lot to solve this pb by myself and I couldn't ..

**Comment by AlexeyAB** [comment]
@ghadahamed Hi,

> Even when I test with the same trained image but rotated by different angle, it is not detected as well. While on the other side when test with any trained image, the mass is detected accurately 

for each object which you want to detect - there must be at least 1 similar object in the Training dataset with about the same: shape, side of object, relative size, angle of rotation, tilt, illumination.

Read this: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> **for each object which you want to detect - there must be at least 1 similar object in the Training dataset with about the same: shape, side of object, relative size, angle of rotation, tilt, illumination.** So desirable that your training dataset include images with objects at diffrent: scales, rotations, lightings, from different sides, on different backgrounds - you should preferably have 2000 different images for each class or more, and you should train 2000*classes iterations or more

----

> I put the test data with the trained data in the data/obj folder as well and write its path in the test.txt file in the data folder.

Did you calculate mAP on test-dataset or on training-dataset?

Try to set `valid=test.txt` in `obj.data`
and run
`./darknet detector map data/obj.data yolo-obj.cfg backup\yolo-obj_best.weights`
what mAP do you get?

**Comment by ghadahamed** [comment]
Okay but this means that we couldn't detect a new mammogram for a new patient as we didn't train the model on a similar mammogram for him. Is this true? and if true this means that the model will get good accuracy with any dataset similar to the trained used one.

I get the map and here it is
![image](https://user-images.githubusercontent.com/47919136/59388727-cd307480-8d6c-11e9-93e5-aaed1c571262.png)

**Comment by AlexeyAB** [comment]
@ghadahamed 

As I see you have good accuracy on Test dataset.

Why do you want to detect on rotated objects? Will be there such rotation in production?

----

> Okay but this means that we couldn't detect a new mammogram for a new patient as we didn't train the model on a similar mammogram for him. Is this true? 

This is true for **any** Detection algorithm.

Very simplified - Yolo can detect objects that are differ from training objects not more than +-15% (size, scale, color, aspect ratio, angle, tilt...).
Some of these things are solved by using automatic data augmentation, except rotation angle.

Solution - you should add all possible variants of objects, as much as possible.

**Comment by ghadahamed** [comment]
I did this to augment data in order to increase its size so help in training different views for mammogram. Kindly let me know your opinion in this as well @AlexeyAB 

Really I am grateful for you and so sorry for disturbing you. 

I want to ask about something are these results for the data in the valid text file or the test text file as the valid text file containing the same images as determined by the main blog?

**Comment by AlexeyAB** [comment]
If rotated objects will be in the final system - then yes - use both original and rotated objects in training dataset.

> I want to ask about something are these results for the data in the valid text file or the test text file as the valid text file containing the same images as determined by the main blog?

What do you mean? I didn't understand.

The mAP is calculated for dataset that is specified in `valid=` in obj.data file.

Do you have train.txt and test.txt, or do you have train.txt, valid.txt and test.txt?

**Comment by ghadahamed** [comment]
Unfortunately train, valid and test and valid = train 
I now get map for the test set and this is the result :((
![image](https://user-images.githubusercontent.com/47919136/59391872-f7d2fb00-8d75-11e9-8c8e-a3635e27a15e.png)

I did something wrong :( So sorry for bothering you @AlexeyAB  :(

**Comment by AlexeyAB** [comment]
@ghadahamed 

Do you get new images for test-dataset?
Can you show images from train and test dataset, what is the difference?
Do you use only rotated images in test-dataset, and only original images in train-dataset?

**Comment by ghadahamed** [comment]
- Yes, I divide the mammograms in the same dataset into 80% for training and 20% for the test.
- Yes, sure. Here is both [https://www.dropbox.com/sh/ocuepri31z80s9u/AAAwlSxvhrXMLm3scTrodaZ6a?dl=0]
- No, I didn't use rotated images neither in the training or the test set, I just when no detection occurs I tried for test one from the training test but rotated. While the main train and test dataset are not rotated.

**Comment by AlexeyAB** [comment]
@ghadahamed 

Did you check that your dataset is correct?
Try to run training with flag ` -show_imgs` in the training command.

Is this label in your dataset correct?
![image](https://user-images.githubusercontent.com/4096485/59394046-e2b29800-8d86-11e9-8576-24d4ecc60ebc.png)

**Comment by ghadahamed** [comment]
The dataset is the INbreast dataset, it is labelled in XML so I map this XML to the data format YOLO needs as in the obj folder as follows:
1- Read the coordinates of the masses in a mammograms from xml and this is the output coordinates (x_min, y_min, x_max, y_max) in this file: https://www.dropbox.com/s/amds1072ton1kpm/detectannotmass.txt?dl=0
2- Generate Yolo annotations by normalizing the coordinates to be within 0 & 1 as explained in the blog
INPUT: x_min, y_min, x_max, y_max
Output for each mass:
x_yolo = (Math.Abs(x_max + x_min) / 2)/w;
y_yolo = (Math.Abs(y_max + y_min) / 2)/h;
width_yolo = Math.Abs(x_max - x_min) / w;
height_yolo = Math.Abs(y_max - y_min) / h;
string class_type = "benign";
if (mammo_class == 1)
    class_type = "malignant";
to generate a file with mass annotations for each mammogram (that are placed in the obj folder)
==================================================
I trained with flag -show_imgs in the training command, this is output, it's weird somehow
![image](https://user-images.githubusercontent.com/47919136/59394403-dc6ced80-8d7f-11e9-841d-713ebb603169.png)
===================================================
Yes it is correct. These are the label names I wrote in obj.names file as follows in the same format:
benign_mass
malignant_mass

**Comment by AlexeyAB** [comment]
This is strange. Why is the label located outside the breast?

![image](https://user-images.githubusercontent.com/4096485/59394046-e2b29800-8d86-11e9-8576-24d4ecc60ebc.png)

**Comment by ghadahamed** [comment]
@AlexeyAB first of all I want to deeply thank you for your great and continuous support. Finally I got results for the test dataset. The problem as you guided me in the labelling of the dataset. I updated all the data, retrain and then test and finally results are obtained.

Thanks a lot @AlexeyAB and I am so sorry for disturbing you

**Comment by mariojy10** [comment]
Hi @ghadahamed this is rather a old post but i hope this reach you. I working with the same dataset the INbreast and i would like to know how you know the severity i.e the malignant and benign , because as far i know the dataset only provide the BI-Rads category. Thank you
